tools/cosim.py

```diff
@@ -0,0 +1,113 @@
+import argparse
+import re
+from pathlib import Path
+
+kPreambuleRegx = r"[-]+"
+kHexRegx = r"[0-9A-Fa-f]+"
+kInstrNumRegx = r"NUM=({})".format(kHexRegx)
+kPCRegx = r"PC=0x({})".format(kHexRegx)
+kRegAccessRegx = r"x({})=0x({})".format(kHexRegx, kHexRegx)
+kMemAccessRegx = r"M\[0x({})\]=0x({})".format(kHexRegx, kHexRegx)
+
+class State:
+    def __init__(self, stream):
+        self.stream = stream
+        self.stream.readline() # skip first preambula in trace. TODO: regex check
+
+        self.currentInstrCount = 0
+        self.PC = 0
+        self.lastChangedRegNum = None
+        self.lastChangedRegVal = None
+        self.lastChangedMemAddr = None
+        self.lastChangedMemVal = None
+        self.EOF = False
+
+    def update(self):
+        finishState = self.EOF
+        while(not finishState):
+            line = self.stream.readline()
+            if not line:
+                finishState = True
+                self.EOF = True
+                break
+
+            match = re.match(kPreambuleRegx, line)
+            if match != None:
+                finishState = True
+                break
+
+            match = re.match(kInstrNumRegx, line)
+            if match != None:
+                self.currentInstrCount = int(match.group(1))
+                break
+
+            match = re.match(kPCRegx, line)
+            if match != None:
+                self.PC = int(match.group(1))
+                break
+
+            match = re.match(kRegAccessRegx, line)
+            if match != None:
+                self.lastChangedRegNum = int(match.group(1))
+                self.lastChangedRegVal = int(match.group(2), base = 16)
+                break
+
+            match = re.match(kMemAccessRegx, line)
+            if match != None:
+                self.lastChangedMemAddr = int(match.group(1), base = 16)
+                self.lastChangedMemVal = int(match.group(2), base = 16)
+                break
+
+class Cosim:
+    def __init__(self, masterFile, slaveFile):
+        self.master = masterFile
+        self.slave = slaveFile
+        self.errorMsg = ""
+        self.isSame = True
+
+    def run(self):
+        masterState = State(self.master)
+        slaveState = State(self.slave)
+
+        while(self.isSame and not (masterState.EOF and slaveState.EOF)):
+            masterState.update()
```

Comment:
Here you can ![open two files](https://stackoverflow.com/questions/4617034/how-can-i-open-multiple-files-using-with-open-in-python) via `with` statement and work with them